File: reference_code/quickstart.py
import io
import json
import os
import sys
import tarfile
import time

import requests  # pip install requests

# NOTE: Provide your API Key here
API_KEY = os.getenv("SINDRI_API_KEY", "")

API_VERSION = "v1"
API_URL = f"https://forge.sindri.app/api/{API_VERSION}/"

api_key_querystring = f"?api_key={API_KEY}"
headers_json = {
    "Accept": "application/json",
    "Authorization": f"Bearer {API_KEY}"
}

circuit_upload_path = os.path.abspath(os.path.join("..", "circuit_database", "circom", "multiplier2"))

# Create a tar archive of the circuit and upload via byte stream
fh = io.BytesIO()
with tarfile.open(fileobj=fh, mode='w:gz') as tar:
    tar.add(circuit_upload_path, arcname="upload.tar.gz")
files = {"files": fh.getvalue()}

# Create new circuit
print("1. Creating circuit...")
response = requests.post(
    API_URL + "circuit",
    headers=headers_json,
    data={"circuit_name": "multiplier2"},
    files=files,
)
assert response.status_code == 201, f"Expected status code 201, received {response.status_code}."
circuit_id = response.json().get("circuit_id")  # Obtain circuit_id


# Poll circuit detail unitl it has a status of Ready or Failed
TIMEOUT = 600  # timeout after 10 minutes
for i in range(TIMEOUT):
    response = requests.get(
        API_URL + f"circuit/{circuit_id}/detail",
        headers=headers_json,
        params={"include_verification_key": False},
    )
    assert (
        response.status_code == 200
    ), f"Expected status code 200, received {response.status_code}."
    status = response.json()["status"]
    if status in ["Ready", "Failed"]:
        print(f"Circuit poll exited after {i} seconds with status: {status}")
        break
    time.sleep(1)
else:
    sys.exit("Circuit compile polling timed out")

# Check for compilation issues
if status == "Failed":
    sys.exit("Circuit compilation failed")
else:
    print("Circuit compilation succeeded!")

# Initiate proof generation
print("2. Proving circuit...")
proof_input = json.dumps({"a": "7", "b": "42"})
response = requests.post(
    API_URL + f"circuit/{circuit_id}/prove",
    headers=headers_json,
    data={"proof_input": proof_input},
)
assert response.status_code == 201, f"Expected status code 201, received {response.status_code}."
proof_id = response.json()["proof_id"]  # Obtain proof_id

# Poll proof detail unitl it has a status of Ready or Failed
TIMEOUT = 1200  # timeout after 20 minutes
for i in range(TIMEOUT):
    response = requests.get(
        API_URL + f"proof/{proof_id}/detail",
        headers=headers_json,
        params={
            "include_proof_input": False,
            "include_public": True,
            "include_verification_key": False,
            "include_proof": False,
        },
    )
    assert (
        response.status_code == 200
    ), f"Expected status code 200, received {response.status_code}."
    status = response.json()["status"]
    if status in ["Ready", "Failed"]:
        print(f"Proof poll exited after {i} seconds with status: {status}")
        break
    time.sleep(1)
else:
    sys.exit("Proof polling timed out")

# Check for proving issues
if status == "Failed":
    sys.exit("Proving failed")
else:
    # Retrieve output from the proof
    public_output = response.json()["public"]
    print(f"Circuit proof output signal: {public_output[0]}")

File: reference_code/quickstart.js
const fs = require("fs");
const path = require("path");
const process = require("process");

process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0'; // REMOVE AFTER LOCAL DEV TESTS

// NOTE: Install dependencies with `npm i axios form-data tar`.
const axios = require("axios");
const FormData = require("form-data");
const tar = require("tar");

// NOTE: Provide your API key here.
const API_KEY = process.env.SINDRI_API_KEY || "";

const API_VERSION = "v1";
const API_URL = `https://localhost/api/${API_VERSION}`;

const apiKeyQueryString = `?api_key=${API_KEY}`;
const headersJson  = {
  Accept: "application/json",
  'Content-Type': "application/x-www-form-urlencoded",
};

// Utility to poll a detail API endpoint until the status is `Ready` or `Failed`.
// Returns the response object of the final request or throws an error if the timeout is reached.
async function pollForStatus(endpoint, timeout = 20 * 60) {
  for (let i = 0; i < timeout; i++) {
    const response = await axios.get(API_URL + endpoint + apiKeyQueryString, {
      headers: headersJson,
      validateStatus: (status) => status === 200,
    });

    const status = response.data.status;
    if (["Ready", "Failed"].includes(status)) {
      console.log(`Poll exited after ${i} seconds with status: ${status}`);
      return response;
    }

    await new Promise((r) => setTimeout(r, 1000));
  }

  throw new Error(`Polling timed out after ${timeout} seconds.`);
}

async function main() {
  try {
    //Load the circuit's `tar.gz` file.
    const circuitFilePath = path.join(
      __dirname,
      "..",
      "circuit_database",
      "circom",
    );


    const uploadFormData = new FormData();
    uploadFormData.append(
      "files",
      tar.c({ gzip: true, sync: true, cwd: circuitFilePath}, ["multiplier2/"]).read(),
      {
        filename: "upload.tar.gz",
      },
    );
    uploadFormData.append("circuit_name", "multiplier2");

    // Create a new circuit.
    console.log("1. Creating circuit...");
    const createResponse = await axios.post(
      API_URL + "/circuit/create" + apiKeyQueryString,
      uploadFormData,
      {validateStatus: (status) => status === 201},
    );

    const circuitId = createResponse.data.circuit_id;

    // Poll the circuit detail endpoint until the compilation status is `Ready` or `Failed`.
    const {
      data: { status: compileStatus },
    } = await pollForStatus(`/circuit/${circuitId}/detail`);

    // Check for compilation issues.
    if (compileStatus === "Failed") {
      throw new Error("Circuit compilation failed.");
    }
    console.log("Circuit compilation succeeded!");

    // Initiate proof generation.
    console.log("2. Proving circuit...");
    const proofInput = JSON.stringify({ a: "7", b: "42" });
    const proveResponse = await axios.post(
      API_URL + `/circuit/${circuitId}/prove` + apiKeyQueryString,
      { proof_input: proofInput },
      { headers: headersJson, validateStatus: (status) => status === 201 },
    );
    const proofId = proveResponse.data.proof_id;

    // Poll the proof detail endpoint until the compilation status is `Ready` or `Failed`.
    const proofDetailResponse = await pollForStatus(`/proof/${proofId}/detail`);

    // Check for proving issues.
    const proofDetailStatus = proveResponse.data.status;
    if (proofDetailStatus === "Failed") {
      throw new Error("Proving failed");
    }

    // Retrieve output from the proof.
    const publicOutput = proofDetailResponse.data.public[0];
    console.log(`Circuit proof output signal: ${publicOutput}`);
  } catch (error) {
    if (error instanceof Error) {
      console.error(error.message);
    } else {
      console.error("An unknown error occurred.");
    }
  }
}

if (require.main === module) {
  main();
}
